### ISSUE TITLE
Mask_RCNN Backbone from Resnet 101 to Resnet152 problem

### ISSUE THREAD

**Issue by enoceanwei** [matterport/Mask_RCNN#1081]
Dear All,
   I am suffering problem, I try to modify the CNN backbone from default resnet101 to 152 architecture. However, follow the instruction of the resnet paper,  compared to resnet101, I can not get the good or expected result.

in <mrcnn/config.py> file, I changed the backbone name
$ BACKBONE = "resnet152"

in <mrcnn/model.py> file, I add the reset152 graph function.
def resnet152_graph(input_image, architecture, stage5=False, train_bn=True):

in <mrcnn/model.py> file, I revised these follow part:
        if callable(config.BACKBONE):
            _, C2, C3, C4, C5 = config.BACKBONE(input_image, stage5=True,
                                                train_bn=config.TRAIN_BN)
        else:
            # wei changed here, 2018/09/28, for adding a resnet152 network
            # _, C2, C3, C4, C5 = resnet_graph(input_image, config.BACKBONE,
            #                                  stage5=True, train_bn=config.TRAIN_BN)
            _, C2, C3, C4, C5 = resnet152_graph(input_image, config.BACKBONE,
                                                stage5=True, train_bn=config.TRAIN_BN)

  However, compare to the resnet101, the deeper network of resnet152 can not perform the expected result, in other words the mAP is very low and predict results are terrible. Could anybody can help me how to fix the problem.

Many thanks.

Kind regards

Wei

**Comment by simonhandsome** [comment]
would you training the mask rcnn with resnet-152 backbone model to tune its weight from begining to end instead of use the original .h5 file in matterport github of weight to continuing training?

**Comment by enoceanwei** [comment]
@simonhandsome 

First of all,  thank you for your reply.

I have no idea about the how to use the customized for 'resnet152 .h5' file to instead of the original  pre-trained COCO weights (mask_rcnn_coco.h5), to be fare, I am not mind that I need to change the pre-train model for resnet152 training, could you tell me more details about it and how to do that, many thanks?

Best Wishes

Wei

**Comment by simonhandsome** [comment]
that coco h5 model seems to be for use of resnet101 or resnet50 model to continue training and object detection, if you want to have an actual totally resnet152 h5 file, you have to modify the original resnet architecture and train tune weight from begining to end your self and ignore the coco h5 file

Best Regards 
By Simon

**Comment by enoceanwei** [comment]
@simonhandsome

Thank you for your info again, there is still a couple of questions by following your above replies.

1. I am sure that I have changed the resnet152 architecture by the resnet paper,  the brief details similar like below: 
"
in <mrcnn/config.py> file, I changed the backbone name
$ BACKBONE = "resnet152"

in <mrcnn/model.py> file, I add the reset152 graph function.
def resnet152_graph(input_image, architecture, stage5=False, train_bn=True):

in <mrcnn/model.py> file, I revised these follow part:
if callable(config.BACKBONE):
_, C2, C3, C4, C5 = config.BACKBONE(input_image, stage5=True,
train_bn=config.TRAIN_BN)
else:
_, C2, C3, C4, C5 = resnet152_graph(input_image, config.BACKBONE,
stage5=True, train_bn=config.TRAIN_BN)
"
2. Follow your opinion, the release file 'mask_rcnn_coco.h5' was trained by coco datasets which are training by default resnet101 backbone in the matterport maskrcnn source code, I understand the right?
    If I understand it correctly, I think I have two ways to train it. 
    First of all, I can train a new model on coco datasets from beginning to end by resnet152 architecture. (Problem: My machine is laptop, I worried about that training the coco datasets from begining is impossible.)
    Secondly, I can find a  .h5 files, which had been trained on resnet152 backbone by someone else. 
(Problem: I can not find it right now)

    However, both of above methods still need to load the weights in the source code <samples-->coco.py>. 
    "
        print("Loading weights ", model_path)
        model.load_weights(model_path, by_name=True)
    "
    My question is that how to avoid to load the pre-train model weight, I just want to train the new resnet152 model from 'scratch' (which is meaning that train from very beginning to end without using the pre-train model)

**Comment by simonhandsome** [comment]
ok, now i clearly know your problem, i cant find the resnet152 trained on coco dataset h5 file either, so you had to train your self from begining to end with resnet152 model, as for the machine resource problem, sorry, i cant help you~

Best Regards~
By Simon

**Comment by enoceanwei** [comment]
Dear Simon,
 
Fist of all, I am very appreciate for your reply. 

To follow your suggestion below:
"
would you training the mask rcnn with resnet-152 backbone model to tune its weight from begining to end instead of use the original .h5 file in matterport github of weight to continuing training?
"

I cancel the below code. Because,  I want to use my modified resnet152 backbeon to train on my own dataset without load the original weight from 'mask_rcnn_coco.h5'. In other words, I want to train my own dataset in a 'scratch' way. 
>  # model.load_weights(model_path, by_name=True)

However, the final result is bad, I use my test datasets to verify the model result, there is no any roi's() can be detected in a test image, I have no idea about these, could you give me some suggestion? many thanks.

Kind regards

Wei

**Comment by ChauncyFr** [comment]
I am also trying to use ResNet152, can you provide your modified resnet152 function? Thank you @enoceanwei

**Comment by MalusJ** [comment]
` input = tf.zeros((kwargs['images_per_gpu'], kwargs['image_shape'][0], kwargs['image_shape'][0], 3), dtype=tf.float32)
net.backbone(input)
net.backbone.load_weights('*.h5', by_name=True)`

here is my code how I swtiched the backbone. I suppose this works.

**Comment by pp-radeep** [comment]
> I am also trying to use ResNet152, can you provide your modified resnet152 function? Thank you @enoceanwei

*********************************************************************************************************************************
def resnet_graph(input_image, architecture, stage5=False, train_bn=True):
    assert architecture in ["resnet50", "resnet101"]
    # Stage 1
    x = KL.ZeroPadding2D((3, 3))(input_image)
    x = KL.Conv2D(64, (7, 7), strides=(2, 2), name='conv1', use_bias=True)(x)
    x = BatchNorm(name='bn_conv1')(x, training=train_bn)
    x = KL.Activation('relu')(x)
    C1 = x = KL.MaxPooling2D((3, 3), strides=(2, 2), padding="same")(x)
    # Stage 2
    x = conv_block(x, 3, [64, 64, 256], stage=2, block='a', strides=(1, 1), train_bn=train_bn)
    x = identity_block(x, 3, [64, 64, 256], stage=2, block='b', train_bn=train_bn)
    C2 = x = identity_block(x, 3, [64, 64, 256], stage=2, block='c', train_bn=train_bn)
    # Stage 3
    x = conv_block(x, 3, [128, 128, 512], stage=3, block='a', train_bn=train_bn)
    x = identity_block(x, 3, [128, 128, 512], stage=3, block='b', train_bn=train_bn)
    x = identity_block(x, 3, [128, 128, 512], stage=3, block='c', train_bn=train_bn)
    x = identity_block(x, 3, [128, 128, 512], stage=3, block='d', train_bn=train_bn)
    x = identity_block(x, 3, [128, 128, 512], stage=3, block='e', train_bn=train_bn)
    x = identity_block(x, 3, [128, 128, 512], stage=3, block='f', train_bn=train_bn)
    x = identity_block(x, 3, [128, 128, 512], stage=3, block='g', train_bn=train_bn)
    C3 = x = identity_block(x, 3, [128, 128, 512], stage=3, block='h', train_bn=train_bn)
    # Stage 4
    x = conv_block(x, 3, [256, 256, 1024], stage=4, block='a', train_bn=train_bn)
    for i in range(35):
        x = identity_block(x, 3, [256, 256, 1024], stage=4, block='b' + str(i), train_bn=train_bn)
    C4 = x
    # Stage 5
    if stage5:
        x = conv_block(x, 3, [512, 512, 2048], stage=5, block='a', train_bn=train_bn)
        x = identity_block(x, 3, [512, 512, 2048], stage=5, block='b', train_bn=train_bn)
        C5 = x = identity_block(x, 3, [512, 512, 2048], stage=5, block='c', train_bn=train_bn)
    else:
        C5 = None
    return [C1, C2, C3, C4, C5]
********************************************************************************************************************************
Never mind what architecture you specify in the config file. B'coz this is not gonna use it.

**Comment by gg123xy** [comment]
i have a little problem, in the code, c1=x,c2=x,c3=x...,if in this way, c1,c2 and c3 will have same values, any error?
